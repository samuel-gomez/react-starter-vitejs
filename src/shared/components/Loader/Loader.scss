@import '@axa-fr/react-toolkit-core/src/common/scss/core.scss';
@import '../../scss/variables.scss';

$base-color: #eaeaf8;
$shine-color: #ffffff;
$classBase: af-loader;

.#{$classBase} {
  position: relative;
  width: 100%;
  min-height: calc(100vh - (calc(var(--fullheaderHeight) + var(--footerHeight))));

  &__spinner {
    display: flex;
    justify-content: center;
    align-items: center;
    color: $color-scorpion;
    font-size: 1.1rem;
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 15;
    padding-top: 3rem;
    &::before {
      content: '';
      background: rgba(220, 220, 255, 0.75);
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
      position: absolute;
      z-index: -1;
    }
    &::after {
      content: '';
      position: absolute;
      left: 50%;
      top: calc(50% - 1rem);
      transform: translate(-50%, -50%);
      width: 50px;
      height: 50px;
      background-color: $brand-primary;
      animation: rotateplane 1.1s infinite ease-in-out;
    }
  }
  &--skeleton {
    min-height: inherit;
  }
  &--skeleton &__spinner {
    padding: 0;
    font-size: 0;
    &::before {
      background: linear-gradient(-45deg, $base-color 40%, $shine-color 50%, $base-color 60%);
      background-size: 400% 400%;
      animation: pulse 1s ease-in infinite;
    }
    &::after {
      display: none;
    }
  }
  &--fullscreen &__spinner {
    height: 100vh;
    position: fixed;
  }
}

@keyframes rotateplane {
  0% {
    transform: translate(-50%, -50%) perspective(120px) rotateX(0deg) rotateY(0deg);
  }

  50% {
    transform: translate(-50%, -50%) perspective(120px) rotateX(-180.1deg) rotateY(0deg);
  }

  100% {
    transform: translate(-50%, -50%) perspective(120px) rotateX(-180deg) rotateY(-179.9deg);
  }
}

@keyframes pulse {
  0% {
    background-position: 0% 0%;
  }
  100% {
    background-position: -100% 0%;
  }
}
