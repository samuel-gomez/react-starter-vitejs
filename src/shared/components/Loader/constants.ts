export enum MODES {
  none = 'none',
  get = 'get',
  post = 'post',
  delete = 'delete',
  update = 'update',
  error = 'error',
}

export enum TEXTS {
  none = '',
  get = 'Chargement en cours',
  post = 'Sauvegarde en cours',
  delete = 'Suppression en cours',
  update = 'Mise Ã  jour en cours',
  error = 'Une erreur est survenue lors du chargement du composant',
}
