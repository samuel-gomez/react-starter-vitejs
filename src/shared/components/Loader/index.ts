export { default } from './Loader.container';
export { TEXTS, MODES } from './constants';
export { default as setLoaderMode } from './setLoaderMode';
export type { TLoaderContainer } from './Loader.container';
