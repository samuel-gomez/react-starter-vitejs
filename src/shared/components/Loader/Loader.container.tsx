import { ComponentProps } from 'react';
import { MODES, TEXTS } from './constants';
import Loader from './Loader';

export type TLoaderContainer = Omit<ComponentProps<typeof Loader>, 'isVisible' | 'message'> & {
  classModifier?: string;
  text?: string;
  mode: keyof typeof MODES;
  LoaderCmpt?: typeof Loader;
};

const LoaderContainer = ({ classModifier = '', text = '', mode = MODES.none, LoaderCmpt = Loader, ...rest }: TLoaderContainer) => {
  const isVisible = mode !== MODES.none;
  const newClassModifier = isVisible ? `${classModifier} active` : classModifier;
  return <LoaderCmpt {...rest} classModifier={newClassModifier} isVisible={isVisible} message={text || TEXTS[mode]} />;
};

export default LoaderContainer;
