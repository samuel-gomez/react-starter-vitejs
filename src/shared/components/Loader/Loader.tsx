import { withClassDefault, withClassModifier, WithClassModifierOptions, compose, identity } from '@axa-fr/react-toolkit-core';
import { ReactNode } from 'react';
import './Loader.scss';

type TLoader = WithClassModifierOptions & {
  className?: string;
  isVisible?: boolean;
  message?: string;
  children?: ReactNode;
};

const Loader = ({ className, isVisible, message, children }: TLoader) => (
  <div className={className}>{isVisible ? <p className="af-loader__spinner">{message}</p> : children}</div>
);

const enhance = compose(identity<TLoader>(), withClassDefault('af-loader'), withClassModifier());
const Enhanced = enhance(Loader);
Enhanced.displayName = 'Skeleton';

export default Enhanced;
