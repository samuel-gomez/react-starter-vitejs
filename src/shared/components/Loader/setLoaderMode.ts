import { MODES } from 'shared/components/Loader';

type TsetLoaderMode = {
  isLoading: boolean;
  LoaderModes?: typeof MODES;
};
const setLoaderMode = ({ isLoading, LoaderModes = MODES }: TsetLoaderMode) => (isLoading ? LoaderModes.get : LoaderModes.none);

export default setLoaderMode;
