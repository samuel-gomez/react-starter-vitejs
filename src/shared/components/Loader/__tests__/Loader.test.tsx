import { describe, it, expect } from 'vitest';
import { render } from '@testing-library/react';
import Loader from '../Loader';

const defaultProps = {
  classModifier: '',
  text: '',
  mode: 'none',
  isVisible: false,
  children: <span>my component to load</span>,
};

describe('<Loader />', () => {
  it('Should render Loader when loading', () => {
    const { asFragment } = render(<Loader {...defaultProps} isVisible />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render Loader when loaded', () => {
    const { asFragment } = render(<Loader {...defaultProps} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
