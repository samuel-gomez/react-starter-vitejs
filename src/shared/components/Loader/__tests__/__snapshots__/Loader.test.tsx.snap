// Vitest Snapshot v1

exports[`<Loader /> > Should render Loader when loaded 1`] = `
<DocumentFragment>
  <div
    class="af-loader"
  >
    <span>
      my component to load
    </span>
  </div>
</DocumentFragment>
`;

exports[`<Loader /> > Should render Loader when loading 1`] = `
<DocumentFragment>
  <div
    class="af-loader"
  >
    <p
      class="af-loader__spinner"
    />
  </div>
</DocumentFragment>
`;
